src/clr-angular/data/tree-view/models/tree-node-model.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { ClrSelectedState } from './selected-state';
+import { BehaviorSubject } from 'rxjs';
+
+export abstract class TreeNodeModel<T> {
+  selected = new BehaviorSubject<ClrSelectedState>(ClrSelectedState.UNSELECTED);
+  model: T | null;
+  // For some reason the polymorphic this doesn't work for either of my use cases. Dammit, Typescript.
+  // So I'm resorting to forcing override with more precise types by marking these abstract.
+  abstract parent: TreeNodeModel<T> | null;
+  abstract children: TreeNodeModel<T>[];
+
+  loading = false;
+
+  destroy() {
+    // Just to be safe
+    this.selected.complete();
+  }
+
+  // Propagate by default when smart, don't propagate in the lazy-loaded tree.
+  setSelected(state: ClrSelectedState, propagateUp: boolean, propagateDown: boolean) {
+    if (state === this.selected.value) {
+      return;
+    }
+    this.selected.next(state);
+    if (propagateDown && state !== ClrSelectedState.INDETERMINATE && this.children) {
+      this.children.forEach(child => child.setSelected(state, false, true));
+    }
+    if (propagateUp && this.parent) {
+      this.parent._updateSelectionFromChildren();
+    }
+  }
+
+  toggleSelection(propagate: boolean) {
+    // NOTE: we always propagate selection up in this method because it is only called when the user takes an action.
+    // It should never be called from lifecycle hooks or app-provided inputs.
+    switch (this.selected.value) {
```

Comment:
Just preference, but seems like easier to do an `if/else` here since there are functionally only two cases.